README.md
# Dos64stb
small stub that allows to run "bare" 64-bit PE binaries in DOS.

details: https://github.com/Baron-von-Riedesel/Dos64-stub/blob/master/src/Readme.txt


History.txt

   18.12.2020:
   - printf implementation made more readable.

   24.10.2020:
   - dos64stb.asm: CR3 was set rather early and assumed that it won't
     be changed in DOS/XMS - which isn't always true (i.e.: HimemSX is
     loaded and the disk cache program uses extended memory beyond 4 GB).

   20.3.2020:
   - IRQ 0+1 (timer+keyboard) now routed to real-mode, in both Dos64stb
     and Dos64stb2.

   12.3.2020:
   - Dos64stb2: IRQ 1 (keyboard) now routed to real-mode.
   - bugfix in JWasm, fixups in 64-bit code will be relative if the target
     segment is 16/32-bit. This allows 64-bit code to directly address 
     variables in 16/32-bit segments.

   17.2.2020:
   - created a second variant of Dos64stb, name Dos64stb2.

   9.2.2020:
   - paging tables and IDT located now before the image, so a heap
     may be initialized above the stack.
   - paging tables setup now done in real-mode to minimize interrupt
     disabled time.
   - keyboard read when interrupt request of keyboard so it won't stop.
   - added DPMI function ax=203h (set exception vector).
   - ensured that Ctrl-C will do a clean exit.

   5.2.2020:
   - no restriction for value of RSP anymore when int 21h is called.
   - transfer flags bits 0-7 from real-mode DOS to 64-bit.
   - sample in C added (TestC.c).
   - Win64 compatible printf.asm added, used by TestC.

   30.1.2020:
   - carry flag of real-mode DOS call wasn't transfered to 64-bit.

   26.1.2020:
   - removed 32-bit segment, jump directly from 16-bit to 64-bit.
   - reprogram master pic only, IRQ0-7 mapped at INT 78-7Fh.
   - implemented DPMI function 0x300 (simulate real-mode interrupt);
     this function is used by the int 21h emulation.


src/TestC.c

/* small test program to test
 * - dos64stub compatibility with MS C.
 * - the micro-printf implementation in printf.asm
 */

#include <stdio.h>

int main() {

	/*
	 * vararg arguments are expanded to 32-bit if they are integers.
	 * addresses are always expanded to 64-bit.
	 */

	unsigned char uc = 'u';
	unsigned short uw = 12;
	unsigned ud = 123456;
	unsigned long long uq = 123456789012345;
	char sc = 's';
	short sw = -12;
	int sd = -123456;
	long long sq = -123456789012345;

	printf( "values 'u',12,123456 | 's',-12,-123456 (%%u | %%d): %u %u %u | %d %d %d\n", uc, uw, ud, sc, sw, sd );
	printf( "char/short/int, cast to long long, %%u | %%d: %u %u %u | %d %d %d\n", (unsigned long long)uc, (unsigned long long)uw, (unsigned long long)ud, (long long)sc, (long long)sw, (long long)sd );
	printf( "char/short/int, cast to long long, %%lu | %%ld: %lu %lu %lu | %ld %ld %ld\n", (unsigned long long)uc, (unsigned long long)uw, (unsigned long long)ud, (long long)sc, (long long)sw, (long long)sd );
	printf( "char/short/int, cast to long long, %%llu | %%lld: %llu %llu %llu | %lld %lld %lld\n", (unsigned long long)uc, (unsigned long long)uw, (unsigned long long)ud, (long long)sc, (long long)sw, (long long)sd );
	printf( "long long, %%u %%d | %%lu %%ld: %u %d | %lu %ld\n", uq, sq, uq, sq );
	printf( "long long, %%llu %%lld: %llu %lld\n", uq, sq );
	printf( "pointers: %p %p %p %p %p %p\n", &uc, &uw, &ud, &sc, &sw, &sd );
	printf( "string: >%s<\n", "this is TestC" );
	return 0;
}


src/MON64.asm

;--- this is a sample showing how a 64-bit binary
;--- may look that can be loaded by dos64stb.bin.

;--- the .x64 directive is not really needed.
;--- it is here so we don't have to use the -win64
;--- cmdline option, which also activates the Win64
;--- calling convention ( not wanted here ).

    .x64
    .model flat
    option casemap:none

lf  equ 10 

;--- define a string
CStr macro text:vararg
local sym
    .const
sym db text,0
    .code
;    lea rax,sym
;    exitm <rax>
    exitm <addr sym>	;works since jwasm v2.13
endm

    .data

qwRsp   dq 0    ;stack pointer
address dq 0    ;next address for d-cmd

;--- keyboard buffer
buffer db 20 dup (0)
lbuffer equ $ - offset buffer

    .code

    include printf.inc

;--- this is the 64-bit entry point.
;--- it's called by the stub with registers:
;--- rbx: image base
;--- rsp: bottom of (reserved) stack

main proc

    invoke printf, CStr(<lf,"Mon64 loaded at %lX, rsp=%lX",lf>), rbx, rsp
    call set_exception_handlers
nextcmd:
    invoke printf, CStr(<"(cmds: a,c,d,q,r,s or x): ">)
    mov ah,1        ;read a key from keyboard with echo
    int 21h
    lea rcx,[nextcmd]
    push rcx
    mov qwRsp, rsp
    push rax
    invoke printf, CStr(lf)
    pop rax
    cmp al,'a'
    jz a_cmd
    cmp al,'c'
    jz c_cmd
    cmp al,'d'
    jz d_cmd
    cmp al,'q'
    jz q_cmd
    cmp al,'r'
    jz r_cmd
    cmp al,'s'
    jz s_cmd
    cmp al,'x'
    jz x_cmd
    cmp al,0dh      ;ENTER?
    jz newline
    invoke printf, CStr(<"unknown cmd: %c",lf>), rax
newline:
    ret
main endp

;--- display control and debug registers

c_cmd proc
    mov rbx,cr0
    mov rcx,cr2
    invoke printf, CStr(<"cr0=%16lX  cr2=%16lX",lf>), rbx, rcx
    mov rbx,cr3
    mov rcx,cr4
    invoke printf, CStr(<"cr3=%16lX  cr4=%16lX",lf>), rbx, rcx
    mov rbx,cr8
    invoke printf, CStr(<"cr8=%16lX",lf>), rbx
    mov rbx,dr0
    mov rcx,dr1
    invoke printf, CStr(<"dr0=%16lX  dr1=%16lX",lf>), rbx, rcx
    mov rbx,dr2
    mov rcx,dr3
    invoke printf, CStr(<"dr2=%16lX  dr3=%16lX",lf>), rbx, rcx
    mov rbx,dr6
    mov rcx,dr7
    invoke printf, CStr(<"dr6=%16lX  dr7=%16lX",lf>), rbx, rcx
    ret
c_cmd endp

;--- get a line of characters
;--- rbx->buffer
;--- esi=size of buffer

getline proc
    xor edi, edi
    dec esi
nextkey:
    mov ah,1
    int 21h
    cmp al,0dh
    jz enter_pressed
    cmp al,08h
    jz backspace_pressed
    mov [rbx+rdi],al
    inc edi
    cmp edi,esi
    jnz nextkey
    mov dl,0dh
    mov ah,2
    int 21h
    jmp enter_pressed
backspace_pressed:
    mov dl,20h
    mov ah,2
    int 21h
    cmp edi,0
    jz nextkey
    dec edi
    mov dl,08h
    mov ah,2
    int 21h
    jmp nextkey
enter_pressed:    
    mov byte ptr [rbx+rdi],0
    ret
getline endp

;--- enter address (for d cmd)

a_cmd proc
    invoke printf, CStr(<"enter start address for d cmd: ">)
    lea rbx, buffer
    mov esi, lbuffer
    call getline
    and edi,edi        ;at least 1 digit entered?
    jz done
    xor edi,edi
    xor esi,esi
    .while byte ptr [rbx+rdi]
        mov al,byte ptr [rbx+rdi]
        sub al,'0'
        jc error
        cmp al,9
        jbe @F
        sub al, 27h
        cmp al, 0fh
        ja error
@@:
        movzx rax,al
        shl rsi,4
        add rsi,rax
        inc edi
    .endw
    mov rax,rsi
    shr rax,48
    and rax, rax
    jz @F
    invoke printf, CStr(<"hint: magnitude of address > 48 bits, exceeds paging capacity",lf>)
@@:
    mov address,rsi
done:
    invoke printf, CStr(lf)
    ret
error:
    invoke printf, CStr(<lf,"%s?",lf>), addr buffer
    ret

a_cmd endp

;--- display memory dump

d_cmd proc
    mov rdi,[address]
    mov ch,8
nextline:
    push rcx
    invoke printf, CStr("%012lX: "), rdi
    pop rcx
    mov cl,16
    .while cl
        push rcx
        mov dl,[rdi]
        movzx rdx,dl
        invoke printf, CStr("%02X "), rdx
        pop rcx
        inc rdi
        dec cl
    .endw
    mov dl,' '
    mov ah,2
    int 21h
    sub rdi,16
    mov cl,16
    .while cl
        mov dl,[rdi]
        cmp dl,20h
        jnc @F
        mov dl,'.'
@@:
        mov ah,2
        int 21h
        inc rdi
        dec cl
    .endw
    push rcx
    invoke printf,CStr(lf)
    pop rcx
    dec ch
    jnz nextline
    mov [address],rdi
    ret
d_cmd endp

;--- display standard registers

r_cmd proc
    invoke printf, CStr(<"rax=%16lX  rbx=%16lX",lf>), rax, rbx
    invoke printf, CStr(<"rcx=%16lX  rdx=%16lX",lf>), rcx, rdx
    invoke printf, CStr(<"rsi=%16lX  rdi=%16lX",lf>), rsi, rdi
    invoke printf, CStr(<"rbp=%16lX  rsp=%16lX",lf>), rbp, rsp
    invoke printf, CStr(<" r8=%16lX   r9=%16lX",lf>),  r8,  r9
    invoke printf, CStr(<"r10=%16lX  r11=%16lX",lf>), r10, r11
    invoke printf, CStr(<"r12=%16lX  r13=%16lX",lf>), r12, r13
    invoke printf, CStr(<"r14=%16lX  r15=%16lX",lf>), r14, r15
    pushfq
    pop rax
    invoke printf, CStr(<"flags=%14lX",lf>), rax
    ret
r_cmd endp

;--- display system registers

s_cmd proc
    sub rsp,16
    sgdt [rsp]
    mov rdi, [rsp+2]
    movzx rsi, word ptr [rsp]
    invoke printf, CStr(<"GDTR base=%lX  limit=%X",lf>), rdi, rsi
    inc rsi
    xor ebx,ebx
    .while rbx < rsi
        mov ch, [rdi+7]
        mov cl, [rdi+4]
        shl ecx,16
        mov cx, [rdi+2]
        movzx rdx, word ptr [rdi+5]
        xor eax,eax
        mov al, dh
        and al, 0Fh
        shl eax, 16
        mov ax, [rdi+0]
        test dh,80h
        jz @F
        shl rax,12
        or ax,0fffh
@@:
        mov r8, rax
        invoke printf, CStr(<"%04X: base=%08lX, limit=%08lX, attr=%04X",lf>), rbx, rcx, r8, rdx
        add rdi,8
        add rbx,8
    .endw

    sidt [rsp]
    mov rdi, [rsp+2]
    movzx rsi, word ptr [rsp]
    invoke printf, CStr(<"IDTR base=%lX  limit=%X",lf>), rdi, rsi
    invoke printf, CStr(<" #      addr     attr      addr     attr      addr     attr      addr     attr",lf>)
    xor ebx,ebx
    .while ebx < 64
        test bl,3
        jnz @F
        invoke printf, CStr(<"%2X ">), rbx
@@:
        mov si,[rdi+6]
        shl esi,16
        mov si,[rdi+0]
        movzx ecx,word ptr [rdi+2]
        movzx edx,word ptr [rdi+4]
        invoke printf, CStr(<"%4X:%08X-%4X ">), rcx, rsi, rdx
        inc ebx
        test bl,3
        jnz @F
        invoke printf, CStr(<lf>)
@@:
        add rdi,16
    .endw
    add rsp,16
    ret
s_cmd endp

;--- display xmm registers

x_cmd proc
    sub rsp,32
    movdqu [rsp+00], xmm0
    movdqu [rsp+16], xmm1
    mov rcx,[rsp+0]
    mov rdx,[rsp+8]
    mov rsi,[rsp+16]
    mov rdi,[rsp+24]
    invoke printf, CStr(<" xmm0=%016lX-%016lX  xmm1=%016lX-%016lX",lf>),rdx, rcx, rdi, rsi
    movdqu [rsp+00], xmm2
    movdqu [rsp+16], xmm3
    mov rcx,[rsp+0]
    mov rdx,[rsp+8]
    mov rsi,[rsp+16]
    mov rdi,[rsp+24]
    invoke printf, CStr(<" xmm2=%016lX-%016lX  xmm3=%016lX-%016lX",lf>),rdx, rcx, rdi, rsi
    movdqu [rsp+00], xmm4
    movdqu [rsp+16], xmm5
    mov rcx,[rsp+0]
    mov rdx,[rsp+8]
    mov rsi,[rsp+16]
    mov rdi,[rsp+24]
    invoke printf, CStr(<" xmm4=%016lX-%016lX  xmm5=%016lX-%016lX",lf>),rdx, rcx, rdi, rsi
    movdqu [rsp+00], xmm6
    movdqu [rsp+16], xmm7
    mov rcx,[rsp+0]
    mov rdx,[rsp+8]
    mov rsi,[rsp+16]
    mov rdi,[rsp+24]
    invoke printf, CStr(<" xmm6=%016lX-%016lX  xmm7=%016lX-%016lX",lf>),rdx, rcx, rdi, rsi
    movdqu [rsp+00], xmm8
    movdqu [rsp+16], xmm9
    mov rcx,[rsp+0]
    mov rdx,[rsp+8]
    mov rsi,[rsp+16]
    mov rdi,[rsp+24]
    invoke printf, CStr(<" xmm8=%016lX-%016lX  xmm9=%016lX-%016lX",lf>),rdx, rcx, rdi, rsi
    movdqu [rsp+00], xmm10
    movdqu [rsp+16], xmm11
    mov rcx,[rsp+0]
    mov rdx,[rsp+8]
    mov rsi,[rsp+16]
    mov rdi,[rsp+24]
    invoke printf, CStr(<"xmm10=%016lX-%016lX xmm11=%016lX-%016lX",lf>),rdx, rcx, rdi, rsi
    movdqu [rsp+00], xmm12
    movdqu [rsp+16], xmm13
    mov rcx,[rsp+0]
    mov rdx,[rsp+8]
    mov rsi,[rsp+16]
    mov rdi,[rsp+24]
    invoke printf, CStr(<"xmm12=%016lX-%016lX xmm13=%016lX-%016lX",lf>),rdx, rcx, rdi, rsi
    movdqu [rsp+00], xmm14
    movdqu [rsp+16], xmm15
    mov rcx,[rsp+0]
    mov rdx,[rsp+8]
    mov rsi,[rsp+16]
    mov rdi,[rsp+24]
    invoke printf, CStr(<"xmm14=%016lX-%016lX xmm15=%016lX-%016lX",lf>),rdx, rcx, rdi, rsi
    add rsp,32
    ret
x_cmd endp

;--- 'q': back to real-mode

q_cmd proc
    mov ax,4c00h
    int 21h
q_cmd endp

;--- set exception 0D+0E handlers so we 
;--- won't terminate unexpectedly

set_exception_handlers proc

    mov bl,0Dh
    lea rdx, exception0D
    mov ecx,cs
    mov ax,0203h
    int 31h
    mov bl,0Eh
    lea rdx, exception0E
    int 31h
    ret
set_exception_handlers endp

;--- handle protection and page faults

exception0D:
    sti
    mov rdx,[rsp+0]
    mov rcx,[rsp+8]
    invoke printf, CStr(<lf,"protection fault, errcode=%X rip=%lX",lf>), rdx, rcx
    mov rsp, qwRsp
    ret
exception0E:
    sti
    mov rdx,[rsp+0]
    mov rcx,[rsp+8]
    mov rbx,cr2
    invoke printf, CStr(<lf,"page fault, errcode=%X rip=%lX cr2=%lX",lf>), rdx, rcx, rbx
    mov rsp, qwRsp
    ret

    end main


Samples/printf.asm

;--- simple printf() implementation
;--- understands formats %u,%d,%x,%c,%s,%p

	option proc:private
	option frame:auto
	option win64:1

	.code

;--- ltob(long n, char * s, int base);
;--- convert long to string

ltoa PROC uses rbx rdi number:qword, pBuffer:ptr, base:dword

	mov ch,0
	mov edi, base
	mov rax, number
	cmp edi,-10
	jne @F
	mov edi,10
	and rax,rax
	jns @F
	neg rax
	mov ch,'-'
@@:
	mov rbx,pBuffer
	add rbx,20
	mov BYTE PTR [rbx],0
	dec rbx
@@nextdigit:
	xor rdx, rdx
	div rdi
	add dl,'0'
	cmp dl,'9'
	jbe @F
	add dl,7+20h
@@:
	mov [rbx],dl
	dec rbx
	and rax, rax
	jne @@nextdigit
	cmp ch,0
	je @F
	mov [rbx],ch
	dec rbx
@@:
	inc rbx
	mov rax,rbx
	ret

ltoa ENDP

handle_char proc char:dword

	mov edx,char
	cmp dl,10
	jnz @F
	push rdx
	mov dl,13
	mov ah,2
	int 21h
	pop rdx
@@:
	mov ah,2
	int 21h
	ret

handle_char endp

printf PROC public frame uses rbx rsi rdi fmt:ptr byte, args:VARARG

local flag:byte
local longarg:byte
local size_:dword
local fillchr:dword
local szTmp[24]:byte

	lea rdi,args
@@L335:
	mov rsi,fmt
nextchar:
	lodsb
	or al,al
	je done
	cmp al,'%'
	je formatitem
	invoke handle_char, eax
	jmp nextchar
done:
	xor eax,eax
	ret 

formatitem:
	lea rax,@@L335
	push rax
	xor edx,edx
	mov [longarg],dl
	mov bl,1
	mov cl,' '
	cmp BYTE PTR [rsi],'-'
	jne @F
	dec bl
	inc rsi
@@:
	mov [flag],bl
	cmp BYTE PTR [rsi],'0'
	jne @F
	mov cl,'0'
	inc rsi
@@:
	mov [fillchr],ecx
	mov ebx,edx

	.while ( byte ptr [rsi] >= '0' && byte ptr [rsi] <= '9' )
		lodsb
		sub al,'0'
		movzx eax,al
		imul ecx,ebx,10		;ecx = ebx * 10
		add eax,ecx
		mov ebx,eax
	.endw

	mov [size_],ebx
	cmp byte ptr [rsi],'l'
	jne @F
	mov [longarg],1
	inc rsi
@@:
	lodsb
	mov [fmt],rsi
	cmp al,'x'
	je handle_x
	cmp al,'X'
	je handle_x
	cmp al,'s'
	je handle_s
	cmp al,'u'
	je handle_u
	cmp al,'d'
	je handle_d
	cmp al,'p'
	je handle_p
	cmp al,'c'
	je handle_c
	and al,al
	jnz @F
	pop rax
	jmp done
handle_c:
	mov al,[rdi]
	add rdi,8
@@:
	invoke handle_char, eax
	retn

handle_s:
	mov rsi,[rdi]
	add rdi,8
	jmp print_string

handle_d:
	mov ebx,-10
	jmp handlenum
handle_u:
	mov ebx,10
	jmp handlenum
handle_p:
	mov longarg,1
handle_x:
	mov ebx,16
handlenum:
	cmp [longarg],0
	je @F
	mov rax,[rdi]
	jmp num2str
@@:
	mov eax,[rdi]
	cmp ebx,-10
	jnz @F
	movsxd rax,dword ptr [rdi]
@@:
num2str:
	add rdi,8
	lea rsi,szTmp
	invoke ltoa, rax, rsi, ebx
	mov rsi,rax

print_string:		;print string RSI
	mov rax,rsi
	mov ebx,size_
	.while byte ptr [rsi]
		inc rsi
	.endw
	sub rsi,rax
	xchg rax,rsi
	sub ebx, eax
	.if flag == 1
		.while sdword ptr ebx > 0
			invoke handle_char, fillchr	;print leading filler chars
			dec ebx
		.endw
	.endif

	.while byte ptr [rsi]
		lodsb
		invoke handle_char, eax	;print char of string
	.endw

	
	.while sdword ptr ebx > 0
		invoke handle_char, fillchr	;print trailing spaces
		dec ebx
	.endw
	retn

printf ENDP

	end


Samples/Mon64.asm

;--- this is a sample showing how a 64-bit binary
;--- may look that can be loaded by dos64stb.bin.
;--- needs JWasm v2.14+ (invoke register argument translation).

;--- the .x64 directive is not really needed.
;--- it is here so we don't have to use the -win64
;--- cmdline option, which also activates the Win64
;--- calling convention ( not wanted here ).

    .x64
    .model flat
    option casemap:none
    option proc:private

lf  equ 10 

?V_CMD equ 0	;1=enable 'v' cmd for testing

ifdef ?PE
%   include ?lnkdef		;for option -pe
endif

;--- define a string
CStr macro text:vararg
local sym
    .const
sym db text,0
    .code
;    lea rax,sym
;    exitm <rax>
    exitm <addr sym>	;works since jwasm v2.13
endm

    .data?
qwRsp   dq ?    ;stack pointer
address dq ?    ;next address for d-cmd
;--- keyboard buffer
buffer db 20 dup (?)
lbuffer equ $ - offset buffer

    .code

    include printf.inc

;--- this is the 64-bit entry point.
;--- it's called by the stub with registers:
;--- rbx: image base
;--- rsp: bottom of (reserved) stack

main proc

    invoke printf, CStr(<lf,"Mon64 loaded at %lX, rsp=%lX",lf>), rbx, rsp
    call set_exception_handlers
    mov address,0
nextcmd:
    invoke printf, CStr(<"(cmds: a,c,d,q,r,s or x): ">)
ifdef ?WAITINPM ;wait in protected-mode ( to test interrupts )
@@:
    mov ax,flat:[41ah]
    cmp ax,flat:[41ch]
    jnz @F
    hlt
    jmp @B
@@:
endif
    mov ah,1        ;read a key from keyboard with echo
    int 21h
    lea rcx,[nextcmd]
    push rcx
    mov qwRsp, rsp
    push rax
    invoke printf, CStr(lf)
    pop rax
    cmp al,'a'
    jz a_cmd
    cmp al,'c'
    jz c_cmd
    cmp al,'d'
    jz d_cmd
    cmp al,'q'
    jz q_cmd
    cmp al,'r'
    jz r_cmd
    cmp al,'s'
    jz s_cmd
if ?V_CMD
    cmp al,'v'
    jz v_cmd
endif
    cmp al,'x'
    jz x_cmd
    cmp al,0dh      ;ENTER?
    jz newline
    invoke printf, CStr(<"unknown cmd: %c",lf>), rax
newline:
    ret
main endp

;--- display control and debug registers

c_cmd proc
    mov rbx,cr0
    mov rcx,cr2
    invoke printf, CStr(<"cr0=%16lX  cr2=%16lX",lf>), rbx, rcx
    mov rbx,cr3
    mov rcx,cr4
    invoke printf, CStr(<"cr3=%16lX  cr4=%16lX",lf>), rbx, rcx
    mov rbx,cr8
    invoke printf, CStr(<"cr8=%16lX",lf>), rbx
    mov rbx,dr0
    mov rcx,dr1
    invoke printf, CStr(<"dr0=%16lX  dr1=%16lX",lf>), rbx, rcx
    mov rbx,dr2
    mov rcx,dr3
    invoke printf, CStr(<"dr2=%16lX  dr3=%16lX",lf>), rbx, rcx
    mov rbx,dr6
    mov rcx,dr7
    invoke printf, CStr(<"dr6=%16lX  dr7=%16lX",lf>), rbx, rcx
    ret
c_cmd endp

;--- get a line of characters
;--- rbx->buffer
;--- esi=size of buffer

getline proc
    xor edi, edi
    dec esi
nextkey:
    mov ah,1
    int 21h
    cmp al,0dh
    jz enter_pressed
    cmp al,08h
    jz backspace_pressed
    mov [rbx+rdi],al
    inc edi
    cmp edi,esi
    jnz nextkey
    mov dl,0dh
    mov ah,2
    int 21h
    jmp enter_pressed
backspace_pressed:
    mov dl,20h
    mov ah,2
    int 21h
    cmp edi,0
    jz nextkey
    dec edi
    mov dl,08h
    mov ah,2
    int 21h
    jmp nextkey
enter_pressed:    
    mov byte ptr [rbx+rdi],0
    ret
getline endp

;--- enter address (for d cmd)

a_cmd proc
    invoke printf, CStr(<"enter start address for d cmd: ">)
    lea rbx, buffer
    mov esi, lbuffer
    call getline
    and edi,edi        ;at least 1 digit entered?
    jz done
    xor edi,edi
    xor esi,esi
    .while byte ptr [rbx+rdi]
        mov al,byte ptr [rbx+rdi]
        sub al,'0'
        jc error
        cmp al,9
        jbe @F
        sub al, 27h
        cmp al, 0fh
        ja error
@@:
        movzx rax,al
        shl rsi,4
        add rsi,rax
        inc edi
    .endw
    mov rax,rsi
    shr rax,47
    and rax, rax
    jz @F
    cmp eax,1ffffh
    jz @F
    invoke printf, CStr(<"hint: magnitude of address > 48 bits, exceeds paging capacity">)
@@:
    mov address,rsi
done:
    invoke printf, CStr(lf)
    ret
error:
    invoke printf, CStr(<lf,"%s?",lf>), addr buffer
    ret

a_cmd endp

;--- display memory dump

d_cmd proc
    mov rdi,[address]
    mov ch,8
nextline:
    push rcx
    invoke printf, CStr("%012lX: "), rdi
    pop rcx
    mov cl,16
    .while cl
        push rcx
        mov dl,[rdi]
        movzx rdx,dl
        invoke printf, CStr("%02X "), rdx
        pop rcx
        inc rdi
        dec cl
    .endw
    mov dl,' '
    mov ah,2
    int 21h
    sub rdi,16
    mov cl,16
    .while cl
        mov dl,[rdi]
        cmp dl,20h
        jnc @F
        mov dl,'.'
@@:
        mov ah,2
        int 21h
        inc rdi
        dec cl
    .endw
    push rcx
    invoke printf,CStr(lf)
    pop rcx
    dec ch
    jnz nextline
    mov [address],rdi
    ret
d_cmd endp

;--- display standard registers

r_cmd proc
    invoke printf, CStr(<"rax=%16lX  rbx=%16lX",lf>), rax, rbx
    invoke printf, CStr(<"rcx=%16lX  rdx=%16lX",lf>), rcx, rdx
    invoke printf, CStr(<"rsi=%16lX  rdi=%16lX",lf>), rsi, rdi
    invoke printf, CStr(<"rbp=%16lX  rsp=%16lX",lf>), rbp, rsp
    invoke printf, CStr(<" r8=%16lX   r9=%16lX",lf>),  r8,  r9
    invoke printf, CStr(<"r10=%16lX  r11=%16lX",lf>), r10, r11
    invoke printf, CStr(<"r12=%16lX  r13=%16lX",lf>), r12, r13
    invoke printf, CStr(<"r14=%16lX  r15=%16lX",lf>), r14, r15
    pushfq
    pop rax
    mov ecx,ds
    mov edx,es
    mov ebx,fs
    mov esi,gs
    mov edi,cs
    mov ebp,ss
    invoke printf, CStr(<"flags=%lX cs=%X ss=%X ds=%X es=%X fs=%X gs=%X",lf>), rax, rdi, rbp, rcx, rdx, rbx, rsi
    lsl rax,rdi
    lar rcx,rdi
    shr rcx,8
    lsl rdx,rbp
    lar rbx,rbp
    shr rbx,8
    invoke printf, CStr(<"cslim=%X csattr=%X sslim=%X ssattr=%X",lf>), rax, rcx, rdx, rbx
    ret
r_cmd endp

;--- display system registers

?STARTINT equ 0 ;start of INTs to display

s_cmd proc
    sub rsp,16
    sgdt [rsp]
    mov rdi, [rsp+2]
    movzx rsi, word ptr [rsp]
    sldt eax
    str ecx
    invoke printf, CStr(<"GDTR base=%lX,limit=%X  LDTR=%X  TR=%X",lf>), rdi, rsi, rax, rcx
    inc rsi
    xor ebx,ebx
    .while rbx < rsi
        mov ch, [rdi+7]
        mov cl, [rdi+4]
        shl ecx,16
        mov cx, [rdi+2]
        movzx rdx, word ptr [rdi+5]
        xor eax,eax
        mov al, dh
        and al, 0Fh
        shl eax, 16
        mov ax, [rdi+0]
        test dh,80h
        jz @F
        shl rax,12
        or ax,0fffh
@@:
        invoke printf, CStr(<"%04X: base=%08lX, limit=%08lX, attr=%04X",lf>), rbx, rcx, rax, rdx
        add rdi,8
        add rbx,8
    .endw

    sidt [rsp]
    mov rdi, [rsp+2]
    movzx rsi, word ptr [rsp]
    invoke printf, CStr(<"IDTR base=%lX  limit=%X",lf>), rdi, rsi
    invoke printf, CStr(<" #    addr             attr    addr             attr    addr             attr",lf>)
if ?STARTINT
    add rdi,?STARTINT*16
endif
    lea rsi,[rdi+16*30h]
    mov bh,?STARTINT
    .while rdi < rsi
        movzx eax,bh
        invoke printf, CStr(<"%2X ">), rax
        mov bl,3
nextitem:
        mov rax,[rdi+4]
        mov ax,[rdi+0]
        movzx ecx,word ptr [rdi+2]
        movzx edx,word ptr [rdi+4]
        invoke printf, CStr(<"%2X:%016lX-%4X ">), rcx, rax, rdx
        add rdi,16
        dec bl
        jnz nextitem
        invoke printf, CStr(<lf>)
        add bh,3
    .endw
    add rsp,16
    ret
s_cmd endp

;--- for various testings
if ?V_CMD
v_cmd proc
    mov ax,0400h
    int 31h
    invoke printf, CStr(<"int 31h, ax=0400h:%X",lf>),rax
    ret
v_cmd endp
endif
;--- display xmm registers

x_cmd proc
    sub rsp,32
    movdqu [rsp+00], xmm0
    movdqu [rsp+16], xmm1
    mov rcx,[rsp+0]
    mov rdx,[rsp+8]
    mov rsi,[rsp+16]
    mov rdi,[rsp+24]
    invoke printf, CStr(<" xmm0=%016lX-%016lX  xmm1=%016lX-%016lX",lf>),rdx, rcx, rdi, rsi
    movdqu [rsp+00], xmm2
    movdqu [rsp+16], xmm3
    mov rcx,[rsp+0]
    mov rdx,[rsp+8]
    mov rsi,[rsp+16]
    mov rdi,[rsp+24]
    invoke printf, CStr(<" xmm2=%016lX-%016lX  xmm3=%016lX-%016lX",lf>),rdx, rcx, rdi, rsi
    movdqu [rsp+00], xmm4
    movdqu [rsp+16], xmm5
    mov rcx,[rsp+0]
    mov rdx,[rsp+8]
    mov rsi,[rsp+16]
    mov rdi,[rsp+24]
    invoke printf, CStr(<" xmm4=%016lX-%016lX  xmm5=%016lX-%016lX",lf>),rdx, rcx, rdi, rsi
    movdqu [rsp+00], xmm6
    movdqu [rsp+16], xmm7
    mov rcx,[rsp+0]
    mov rdx,[rsp+8]
    mov rsi,[rsp+16]
    mov rdi,[rsp+24]
    invoke printf, CStr(<" xmm6=%016lX-%016lX  xmm7=%016lX-%016lX",lf>),rdx, rcx, rdi, rsi
    movdqu [rsp+00], xmm8
    movdqu [rsp+16], xmm9
    mov rcx,[rsp+0]
    mov rdx,[rsp+8]
    mov rsi,[rsp+16]
    mov rdi,[rsp+24]
    invoke printf, CStr(<" xmm8=%016lX-%016lX  xmm9=%016lX-%016lX",lf>),rdx, rcx, rdi, rsi
    movdqu [rsp+00], xmm10
    movdqu [rsp+16], xmm11
    mov rcx,[rsp+0]
    mov rdx,[rsp+8]
    mov rsi,[rsp+16]
    mov rdi,[rsp+24]
    invoke printf, CStr(<"xmm10=%016lX-%016lX xmm11=%016lX-%016lX",lf>),rdx, rcx, rdi, rsi
    movdqu [rsp+00], xmm12
    movdqu [rsp+16], xmm13
    mov rcx,[rsp+0]
    mov rdx,[rsp+8]
    mov rsi,[rsp+16]
    mov rdi,[rsp+24]
    invoke printf, CStr(<"xmm12=%016lX-%016lX xmm13=%016lX-%016lX",lf>),rdx, rcx, rdi, rsi
    movdqu [rsp+00], xmm14
    movdqu [rsp+16], xmm15
    mov rcx,[rsp+0]
    mov rdx,[rsp+8]
    mov rsi,[rsp+16]
    mov rdi,[rsp+24]
    invoke printf, CStr(<"xmm14=%016lX-%016lX xmm15=%016lX-%016lX",lf>),rdx, rcx, rdi, rsi
    add rsp,32
    ret
x_cmd endp

;--- 'q': back to real-mode

q_cmd proc
    mov ax,4c00h
    int 21h
q_cmd endp

;--- set exception 0D+0E handlers so we 
;--- won't terminate unexpectedly

set_exception_handlers proc

    mov bl,0Dh
    lea rdx, exception0D
    mov ecx,cs
    mov ax,0203h
    int 31h
    mov bl,0Eh
    lea rdx, exception0E
    int 31h
    ret
set_exception_handlers endp

;--- handle protection and page faults

exception0D:
    sti
    mov rdx,[rsp+0]
    mov rcx,[rsp+8]
    invoke printf, CStr(<lf,"protection fault, errcode=%X rip=%lX",lf>), rdx, rcx
    mov rsp, qwRsp
    ret
exception0E:
    sti
    mov rdx,[rsp+0]
    mov rcx,[rsp+8]
    mov rbx,cr2
    invoke printf, CStr(<lf,"page fault, errcode=%X rip=%lX cr2=%lX",lf>), rdx, rcx, rbx
    mov rsp, qwRsp
    ret

    end main
